Sources/NIOPosix/PipePair.swift

```diff
@@ -13,55 +13,92 @@
+extension SelectablePipeHandle: Selectable {
     func withUnsafeHandle<T>(_ body: (CInt) throws -> T) throws -> T {
-        try self.handle.withUnsafeFileDescriptor(body)
+        try body(self.fileDescriptor)
     }
 }
 
-final class PipePair: SocketProtocol {
-    typealias SelectableType = SelectableFileHandle
-
-    let inputFD: SelectableFileHandle?
-    let outputFD: SelectableFileHandle?
+extension SelectablePipeHandle: CustomStringConvertible {
+    public var description: String {
+        return "SelectableFileHandle(isOpen: \(self.isOpen), fd: \(self.fileDescriptor))"
+    }
+}
 
-    init(inputFD: NIOFileHandle?, outputFD: NIOFileHandle?) throws {
-        self.inputFD = inputFD.flatMap { SelectableFileHandle($0) }
-        self.outputFD = outputFD.flatMap { SelectableFileHandle($0) }
-        try self.ignoreSIGPIPE()
-        for fileHandle in [inputFD, outputFD].compactMap({ $0 }) {
-            try fileHandle.withUnsafeFileDescriptor {
-                try NIOFileHandle.setNonBlocking(fileDescriptor: $0)
+final class PipePair: SocketProtocol {
+    typealias SelectableType = SelectablePipeHandle
+
+    let inputSPH: SelectablePipeHandle?
+    let outputSPH: SelectablePipeHandle?
+
+    init(inputFH: NIOFileHandle?, outputFH: NIOFileHandle?) throws {
+        var errorWeNeedToAbortOn: Error? = nil
+        var inputFD: CInt? = nil
+        var outputFD: CInt? = nil
+        do {
+            inputFD = try inputFH?.takeDescriptorOwnership()
+        } catch {
+            errorWeNeedToAbortOn = error
+        }
+        do {
+            outputFD = try outputFH?.takeDescriptorOwnership()
+        } catch {
+            errorWeNeedToAbortOn = error
+        }
+        if let error = errorWeNeedToAbortOn {
+            if let inputFD = inputFD {
+                try! Posix.close(descriptor: inputFD)
+            }
+            if let outputFD = outputFD {
+                try! Posix.close(descriptor: outputFD)
             }
+            throw error
+        }
+        self.inputSPH = inputFD.flatMap { SelectablePipeHandle(takingOwnershipOfDescriptor: $0) }
+        self.outputSPH = outputFD.flatMap { SelectablePipeHandle(takingOwnershipOfDescriptor: $0) }
+        try self.ignoreSIGPIPE()
+        for fd in [inputFD, outputFD].compactMap({ $0 }) {
+            try NIOFileHandle.setNonBlocking(fileDescriptor: fd)
         }
     }
 
     func ignoreSIGPIPE() throws {
-        for fileHandle in [self.inputFD, self.outputFD].compactMap({ $0 }) {
+        for fileHandle in [self.inputSPH, self.outputSPH].compactMap({ $0 }) {
             try fileHandle.withUnsafeHandle {
                 try PipePair.ignoreSIGPIPE(descriptor: $0)
             }
         }
     }
 
     var description: String {
-        "PipePair { in=\(String(describing: inputFD)), out=\(String(describing: inputFD)) }"
+        "PipePair { in=\(String(describing: self.inputSPH)), out=\(String(describing: self.outputSPH)) }"
```

Comment:
there was actually a bug here (it previously printed the input fd twice, whoops)